source/common/upstream/default_local_address_selector.cc

```diff
@@ -0,0 +1,112 @@
+#include "source/common/upstream/default_local_address_selector.h"
+
+#include <string>
+
+#include "envoy/network/socket.h"
+
+#include "source/common/network/resolver_impl.h"
+#include "source/common/network/socket_option_factory.h"
+
+namespace Envoy {
+namespace Upstream {
+
+namespace {
+constexpr absl::string_view kDefaultLocalAddressSelectorName =
+    "envoy.upstream.local_address_selector.default_local_address_selector";
+} // namespace
+
+DefaultUpstreamLocalAddressSelector::DefaultUpstreamLocalAddressSelector(
+    std::vector<::Envoy::Upstream::UpstreamLocalAddress> upstream_local_addresses,
+    absl::optional<std::string> cluster_name)
+    : upstream_local_addresses_(std::move(upstream_local_addresses)) {
+
+  if (upstream_local_addresses_.size() == 0) {
+    throw EnvoyException(fmt::format("{}'s upstream binding config has no valid source address.",
+                                     !(cluster_name.has_value())
+                                         ? "Bootstrap"
+                                         : fmt::format("Cluster {}", cluster_name.value())));
+  }
+
+  if (upstream_local_addresses_.size() > 2) {
+    throw EnvoyException(fmt::format(
+        "{}'s upstream binding config has more than one extra/additional source addresses. Only "
+        "one extra/additional source can be supported in BindConfig's "
+        "extra_source_addresses/additional_source_addresses field",
+        !(cluster_name.has_value()) ? "Bootstrap"
+                                    : fmt::format("Cluster {}", cluster_name.value())));
+  }
+  // If we have exactly one upstream address, it needs to have a valid IP
+  // version if non-null
+  if (upstream_local_addresses_.size() == 1) {
+    // First verify that all address have valid IP address information.
+    if (upstream_local_addresses_[0].address_ != nullptr &&
+        upstream_local_addresses_[0].address_->ip() == nullptr) {
```

Comment:
We shouldn't have this case, since the API only allow the socket address. It can be just an assertion, but it is fine leave check at here.